### ISSUE TITLE
ZODB 5.6.0 breaks zc.zrs

### ISSUE THREAD

**Issue by dataflake** [zopefoundation/zc.zrs#6]
My setup has a central ZEO server running on ZODB 5.5.1 and a replicated server using `zc.zrs`. Now I updated the replication client to 5.6.0 and the first replicated transaction leads to a disconnection and the replication process breaks down:

```
2020-07-20T04:51:31 listening on /opt/buildout/var/zeoserver.soc
2020-07-20T07:40:51 IPv4Address(type='TCP', host='xxx.xxx.xxx.xxx', port=xxxx): Input data error
Traceback (most recent call last):
  File "/opt/.eggs/zc.zrs-3.1.0-py3.8.egg/zc/zrs/secondary.py", line 82, in dataReceived
    self.__stream(data)
  File "/opt/.eggs/zc.zrs-3.1.0-py3.8.egg/zc/zrs/sizedmessage.py", line 54, in __call__
    self._callback(result)
  File "/opt/.eggs/zc.zrs-3.1.0-py3.8.egg/zc/zrs/secondary.py", line 137, in messageReceived
    self.factory.storage.tpc_begin(
  File "/opt/.eggs/ZODB-5.6.0-py3.8.egg/ZODB/BaseStorage.py", line 193, in tpc_begin
    ext = transaction.extension_bytes
AttributeError: 'Transaction' object has no attribute 'extension_bytes'
```

Downgrading the replication client to ZODB 5.5.1 as workaround fixes it.

**Comment by dataflake** [comment]
@jamadden I have created a branch and draft PR at #7 to get the package into a shape where it's easier to run tests and the tests succeed under Python 3.5, 3.6 and 3.7.

`zc.zrs` has its own rudimentary Transaction class that is being used with the various `tpc_` methods: https://github.com/zopefoundation/zc.zrs/blob/1e0fff450e3ca843861ca3ed519b335e7256e8cb/src/zc/zrs/secondary.py#L307-L318

I haven't worked with this code or ZODB+ZEO code in general so I have a hard time debugging the issue. I have tried substituting the custom Transaction class with `ZODB.Connection.TransactionMetaData` and extending the custom Transaction class so it has the methods/properties provided by the `TransactionMetaData` class, but I'm not really progressing. Debugging is hard because the only thing I see in the tests is methods that fail because the ZEO client disconnects, most likely because the server code errors out, but I cannot see where and how because the tests run the server process in a separate thread and I see no output from it.

**Comment by jamadden** [comment]
Unfortunately, I have never used `zc.zrs` and know next to nothing about it, how it's supposed to work, or how to run its tests (we skipped right past ZEO and onto RelStorage). I do agree that debugging doctests, especially those that spawn separate processes (like the buildout doctests) can be challenging. I sometimes find it more straightforward to port some interesting subset of the doctest into a real unit test; that can make it easier to isolate the set up portion so, for example, one can directly launch the ZEO server in its own terminal with maximum verbosity. 


I would *guess* the easiest way to fix the original issue would be to use `ZODB.Connection.TransactionMetaData` directly. I don't know why the designers didn't do that in the first place (maybe it didn't exist originally?).


Here, it looks like the tests haven't run on CI in 3 years. It looks like they ran last just after ZEO 5.1 was released (though I can't tell what version they ran with). But they were broken even then. Since then there's been ZEO 5.2, so things may have bitrotted a bit out from under `zc.zrs`. Maybe start by trying to use older versions of ZEO/ZODB and gradually bump versions?

**Comment by dataflake** [comment]
The zodb56 branch I created has a `tox` config, so running tests has become a lot easier. The point where things break is when I go from ZODB 5.5.1 to ZODB 5.6.0. The ZEO version doesn't matter.

Like I said, I already tried swapping in `ZODB.Connection.TransactionMetaData` but the failure symptom remains the same - tests fail because somewhere before the ZEO client process has disconnected from the server process, most likely because it hit an exception on the server side.

I'll try to follow your advice and port a failing test to a normal unit test. Using doctests all over was one of the worst decisions ever made by at least some in the Zope developer community.

**Comment by jamadden** [comment]
I just remembered https://github.com/zopefoundation/ZEO/pull/159: It's possible no released version of ZEO correctly works with ZODB 5.6 yet.

**Comment by jamadden** [comment]
Looks like on Python 3 the brokenness manifests as a client disconnection. https://travis-ci.org/github/zopefoundation/ZEO/jobs/706968350#L717  So that seems a likely culprit.

**Comment by dataflake** [comment]
Status update: I'm observing several different breakages see below. At this moment the safest option is to stick with Python 3.7 and ZODB 5.5. Everything higher than that is messed up on multiple levels:

- ZODB 5.6 is not compatible with zc.zrs due to the extension bytes issue. After more testing I can fix that issue by subclassing the custom Transaction class from `ZODB.Connection.TransactionMetaData`, BUT:
- ZODB 5.6 also breaks ZEO. This is fixed for Python up to 3.7 by applying https://github.com/zopefoundation/ZEO/pull/159, BUT:
- Don't upgrade any of this to Python 3.8 because ZEO isn't compatible with that, either. Tracebacks like the following will appear:

```
Error in test test_multiple_addresses (ZEO.asyncio.tests.ClientTests)
Traceback (most recent call last):
  File "/Users/jens/src/Python-3.8.4/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/Users/jens/src/Python-3.8.4/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/Users/jens/src/Python-3.8.4/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/Users/jens/src/zope/ZEO/.tox/py38/lib/python3.8/site-packages/ZEO/asyncio/tests.py", line 427, in test_multiple_addresses
    self.respond(1, None)
  File "/Users/jens/src/zope/ZEO/.tox/py38/lib/python3.8/site-packages/ZEO/asyncio/tests.py", line 116, in respond
    self.loop.protocol.data_received(
  File "/Users/jens/src/zope/ZEO/.tox/py38/lib/python3.8/site-packages/ZEO/asyncio/base.py", line 128, in data_received
    self.message_received(collected)
  File "/Users/jens/src/zope/ZEO/.tox/py38/lib/python3.8/site-packages/ZEO/asyncio/client.py", line 232, in message_received
    future.set_result(args)
  File "/Users/jens/src/zope/ZEO/.tox/py38/lib/python3.8/site-packages/ZEO/asyncio/client.py", line 904, in set_result
    self.cb(self)
  File "/Users/jens/src/zope/ZEO/.tox/py38/lib/python3.8/site-packages/ZEO/asyncio/client.py", line 45, in _
    f = gen.send(result)
  File "/Users/jens/src/zope/ZEO/.tox/py38/lib/python3.8/site-packages/ZEO/asyncio/client.py", line 202, in finish_connect
    self.client.registered(self, server_tid)
  File "/Users/jens/src/zope/ZEO/.tox/py38/lib/python3.8/site-packages/ZEO/asyncio/client.py", line 447, in registered
    self._clear_protocols(protocol)
  File "/Users/jens/src/zope/ZEO/.tox/py38/lib/python3.8/site-packages/ZEO/asyncio/client.py", line 404, in _clear_protocols
    p.close()
  File "/Users/jens/src/zope/ZEO/.tox/py38/lib/python3.8/site-packages/ZEO/asyncio/client.py", line 98, in close
    self._connecting.cancel()
  File "/Users/jens/src/zope/ZEO/.tox/py38/lib/python3.8/site-packages/ZEO/asyncio/testing.py", line 31, in call_soon
    func(*args)
  File "/Users/jens/src/zope/ZEO/.tox/py38/lib/python3.8/site-packages/ZEO/asyncio/client.py", line 129, in done_connecting
    if future.exception() is not None:
asyncio.exceptions.CancelledError
```

**Comment by dataflake** [comment]
Python 3.7 and ZODB 5.5.1 are safe. Neither Python 3.8 nor ZODB 5.6 are, no matter how you combine it.

The production servers I am working with have a primary running Python 3.7 and ZODB 5.5.1. I had upgraded the secondaries to Python 3.8 and ZODB 5.6 and replication broke with the first replicated transaction that reached the secondary. Moving back to ZODB 5.5.1 brought back replication, but the unit tests for ZEO and zc.zrs won't run under either Python 3.8 or ZODB 5.6, so I rolled it all back on the secondaries.

The Zope clients were on Python 3.7 and ZODB 5.6 and we didn't see any issues. But I have changed our buildout configurations so they will roll back to ZODB 5.5.1 on the next update. I don't trust that anymore right now.

**Comment by freddrake** [comment]
Understood; thanks, Jens!

We haven't noticed issues with the Python 3.8 / ZODB 5.6.0 clients, but we're not customer-facing yet.  This would definitely be the time to consider rolling back.

**Comment by dataflake** [comment]
I'm giving up. I don't have the knowledge to fix this. I left the `zodb56` branch around in case anyone wants to take this on and closed the PR.

The tests run fine on macOS but fail on Linux, and there appears to be an issue with blobs.

**Comment by georgpfolz** [comment]
Maybe these informations might prove useful:

- I'm not using blobs (I'm not running Plone) but it fails nonetheless
- As was already mentioned, it only breaks if ZODB 5.6 is on the receiving end: It works with ZODB 5.6 on the primary server and ZODB 5.5.1 on the secondary server

My configuration:
- FreeBSD 12.2 p2
- Python 3.8.9 (I'm not using ZEO, so I'm good)
- Zope 5.2.1

This is an important issue to me, as I would like to be able to use tempstorage again (which was fixed recently but requires ZODB 5.6).
